xla/service/gpu/tests/gpu_fused_mha_test.cc

```diff
@@ -1102,6 +1102,212 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
   }
 };
 
+class FlashAttentionBMMScalePaddingMaskSoftmaxBMMF8
+    : public MultiHeadedAttentionTest {
+ protected:
+  void TestImpl_Flash_Attention_Inference_BMM1_NoMask_Softmax_BMM2_F8() {
+    if (skip_reason_) GTEST_SKIP() << *skip_reason_;
+    if (GetDnnVersionInfoOrDefault(backend().default_stream_executor()) <
+        se::dnn::VersionInfo(9, 1, 0)) {
+      GTEST_SKIP() << "Flash Attention requires cuDNN >= 9.1.0.";
+    }
+    XlaBuilder builder(TestName());
+    std::string hlo_string_ref =
+        R"(
+    HloModule jit__unnamed_wrapped_function_, entry_computation_layout={(bf16[4,16,4,16]{3,2,1,0}, bf16[4,16,4,16]{3,2,1,0}, bf16[4,16,4,16]{3,2,1,0})->bf16[4,4,16,16]{3,1,2,0}}, allow_spmd_sharding_propagation_to_parameters={true,true,true}, allow_spmd_sharding_propagation_to_output={true}
```

Comment:
Can you explain what is the purpose of all the HLO leading to the custom call? Does it actually provide value? If not, should we just compare the upcasted custom call vs the reference one?

Reply:
The HLO leading to the custom-call performs a "[cast-to-representable](https://github.com/google/praxis/blob/619fc3463d0eae7b24b0f4aa508280cba790666e/praxis/layers/injection/fp8_nvidia_gpu_test.py#L41)" operation, which adjusts the input to fit within the range that the FP8 data type can represent. Therefore, it's also necessary for the reference implementation to include this step in order to maintain numerical equivalence.